src/channel/mpmc.rs

```diff
@@ -182,12 +196,27 @@ where
         }
     }
 
+    /// Tries to read the value from the channel without blocking.
+    fn try_receive(&mut self) -> Option<T> {
+        if !self.buffer.is_empty() {
+            let val = self.buffer.pop();
+
+            // Since this means a space in the buffer had been freed,
+            // try to copy a value from a potential waiter into the channel.
+            unsafe { self.try_copy_value_from_last_waiter() };
+
+            Some(val)
+        } else {
```

Comment:
This is missing the support for unbuffered channels (copying over a value directly from a send waiter).

While it seems like `try_send` is problematic for unbuffered channels `try_receive` should be fine.

Would also be great to have some tests for the behavior on unbuffered channels, because their behavior is somewhat special and not covered by the other path.

Reply:
Btw: I think lines https://github.com/Matthias247/futures-intrusive/pull/14/files#diff-3716d1c2cd77fc4e7edcc03980284755R228-R258
would actually be equivalent between both methods and could be moved into a common function.

Reply:
Not following your last comment. Which methods are you referring to?

Reply:
Most parts of the old `try_receive` are equivalent between the new `receive_or_wait` and the new `try_receive`.  They should be highlighted in yellow in the link I pasted.

Reply:
`recv_or_register` now uses `try_receive` internally.

Reply:
That was actually not what I meant. `try_receive` still doesn't handle the case where a `send_waiter` is registered and carries a value, which has to be read. That requires the code from the ` if !self.send_waiters.is_empty()` branch, which is then also equal between the 2 receive path. Therefore I think it might be possible to move some of the logic into a common method (which is neither `try_receive` nor `receive_or_register`) and is called by both.

You want a unit test with an unbuffered channel where one `poll` happens on the `send` Future (which reigsters the `Future` at the channel), and then a `try_receive` on the sender which is expected to complete the transaction.